AppCenter/AppCenter/Internals/Channel/MSChannelGroupDefault.m

```diff
@@ -0,0 +1,190 @@
+  }
+}
+
+- (void)enumerateDelegatesForSelector:(SEL)selector withBlock:(void (^)(id<MSChannelDelegate> delegate))block {
+  @synchronized(self) {
+    for (id<MSChannelDelegate> delegate in self.delegates) {
+      if (delegate && [delegate respondsToSelector:selector]) {
+        block(delegate);
+      }
+    }
+  }
+}
+
+#pragma mark - Channel Delegate
+
+- (void)willSendLog:(id<MSLog>)log {
+  [self enumerateDelegatesForSelector:@selector(willSendLog:)
+                            withBlock:^(id<MSChannelDelegate> delegate) {
+                              [delegate willSendLog:log];
+                            }];
+}
+
+- (void)didSucceedSendingLog:(id<MSLog>)log {
+  [self enumerateDelegatesForSelector:@selector(didSucceedSendingLog:)
+                            withBlock:^(id<MSChannelDelegate> delegate) {
+                              [delegate didSucceedSendingLog:log];
+                            }];
+}
+
+- (void)didFailSendingLog:(id<MSLog>)log withError:(NSError *)error {
+  [self enumerateDelegatesForSelector:@selector(didFailSendingLog:withError:)
+                            withBlock:^(id<MSChannelDelegate> delegate) {
+                              [delegate didFailSendingLog:log withError:error];
+                            }];
+}
+
+- (void)onEnqueuingLog:(id<MSLog>)log withInternalId:(NSString *)internalId {
+  [self enumerateDelegatesForSelector:@selector(onEnqueuingLog:withInternalId:)
+                            withBlock:^(id<MSChannelDelegate> delegate) {
+                              [delegate onEnqueuingLog:log withInternalId:internalId];
+                            }];
+}
+
+- (void)onFinishedPersistingLog:(id<MSLog>)log withInternalId:(NSString *)internalId {
+  [self enumerateDelegatesForSelector:@selector(onFinishedPersistingLog:withInternalId:)
+                            withBlock:^(id<MSChannelDelegate> delegate) {
+                              [delegate onFinishedPersistingLog:log withInternalId:internalId];
+                            }];
+}
+
+- (void)onFailedPersistingLog:(id<MSLog>)log withInternalId:(NSString *)internalId {
+  [self enumerateDelegatesForSelector:@selector(onFailedPersistingLog:withInternalId:)
+                            withBlock:^(id<MSChannelDelegate> delegate) {
+                              [delegate onFailedPersistingLog:log withInternalId:internalId];
+                            }];
+}
+
+#pragma mark - Enable / Disable
+
+- (void)setEnabled:(BOOL)isEnabled andDeleteDataOnDisabled:(BOOL)deleteData {
+
+  // Propagate to sender.
+  [self.sender setEnabled:isEnabled andDeleteDataOnDisabled:deleteData];
+
+  // Propagate to initialized channels.
+  for (id<MSChannelProtocol> channel in self.channels) {
+    [channel setEnabled:isEnabled andDeleteDataOnDisabled:deleteData];
+  }
+
+  /**
+   * TODO: There should be some concept of logs on disk expiring to avoid leaks
+   * when a channel is disabled with lingering logs but never enabled again.
```

Comment:
Reading this, this sounds really bad. Maybe we can add that this really shouldn't happen as we delete when disabling a channel but that the TODO is more about being proactive? Or am I missing something?

Reply:
Added a bit more to this comment